sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala

```diff
@@ -281,6 +281,10 @@ object FileFormatWriter extends Logging {
     } catch {
       case e: FetchFailedException =>
         throw e
+      case f: FileAlreadyExistsException =>
```

Comment:
Looks like this makes a very strong assumption. It assumes `FileAlreadyExistsException` is thrown from the commit protocol. But `FileAlreadyExistsException` is a very general exception which can be thrown from anywhere. What if it's from user codes but the user wants to retry?

Reply:
If users really want to retry, I believe it is to catch `FileAlreadyExistsException` and do retry in user code with changing filename, instead of relying on Spark to retry it.

To rely on Spark's task attempt for file writing retry? It is unreliable and task failure can have many reasons not just file already existing.

Reply:
> If users really want to retry, I believe it is to catch `FileAlreadyExistsException` and do retry in user code with changing filename, instead of relying on Spark to retry it.

This applies to any other exceptions as well. Of cause, users can retry. But since Spark provides task retry, users may have a rare race that can cause `FileAlreadyExistsException` and decide to reply on Spark task retry to handle it. In addition, some users may just call a library or a service that they cannot make changes.

Moreover, this is a behavior change and it may break user applications relying on task retry for `FileAlreadyExistsException`. But it's documented.

IIUC, this PR seems just trying to fail fast when hitting `FileAlreadyExistsException` in some rare case ( SPARK-27194 ). Please correct me if I'm wrong. If so, it's not worth to make a behavior change in my opinion.

Reply:
It is not a rare case actually. We had customer encountered this issue in production jobs occasionally. There are also other people reporting this issue. It is remarkable because this `FileAlreadyExistsException` cannot be handled by user code. It is happened internally in Spark, and users have no way to deal with it. Image your production jobs fail after 10+ hours because of `FileAlreadyExistsException`. I believe it is worth failing fast the production jobs.

It sounds really unreliable to rely on retry to handle a rare race that causes `FileAlreadyExistsException`. What if later attempts cannot success too? Seems to me a production job won't rely on that and needs address this issue seriously. In most of cases this can be handled by catching `FileAlreadyExistsException` and retrying. In rare case if any, there is rare race that can cause `FileAlreadyExistsException`  and it cannot make change? It sounds pretty rare to me, and even if it is really happens we can retry the call of the library/service to solve it, so I'm not sure if we should sacrifice the cases seen in production for a rare case.

Reply:
Makes sense to me. @viirya can we throw a spark specific exception for commit protocol errors and catch it here?

Reply:
@cloud-fan Sure. Will submit a PR.

Reply:
Created SPARK-33167 and will submit a PR in next days. Thanks.

Reply:
I see. Thanks for the details. We have different standpoints. For your cases the first one option looks a better choice. The customers we had are using HDFS and `FileAlreadyExistsException` isn't recoverable. So the pain point comes from more time spent on a failed job.

I believe even SPARK-27194 is resolved, fast-fail of a failed job caused by `FileAlreadyExistsException` or maybe other errors if we know they are un-recoverable in advance, is still useful.

Seems to me there are options, one is to revert this completely, second is to add a config for the fast-fail behavior and set it false by default. I prefer the second one because the reason above, we can relieve the pain of wasting time on failed job if users want.

WDYT?

Reply:
> second is to add a config for the fast-fail behavior and set it false by default. I prefer the second one because the reason above, we can relieve the pain of wasting time on failed job if users want.

+1 for this.